### PR TITLE
PVR DMA implementation

### PR DESCRIPTION
**Pull request by andressbarajas**
### PVR and TA DMA Clarifications and New PVR DMA Functionality
It turns out that what we've been using for submitting textures and vertex data is TA DMA, not PVR DMA as previously thought. This PR adds true PVR DMA functionality to the pvr_dma_*() functions. Below is a breakdown of the capabilities of each DMA type and the changes introduced in this PR.

### DMA Types Overview
=== TA DMA ===
- Can DMA texture data.
- Can DMA vertex data.
- Direction: SH4 => TA only (one-way transfer).

=== PVR DMA ===

- Can DMA texture data.
- Can DMA PVR register data (can be used to DMA palettes and other register data).
- Direction: Bi-directional (SH4 <=> PVR), allowing both upload and download transfers.

### Breaking API Changes
EDIT:
This PR introduces no API-breaking changes but deprecates pvr_dma_ready() and provides 2 new functions to check ready status.

I moved the pvr_dma functions to their own header, pvr_dma.h.  The params to these functions are now more strongly typed with enums rather than ints.

### CODE REVIEW

**Comment by DC-SWAT** [comment]
Have you thought about simply combining them into one API? The choice of DMA can be made based on the `type` of transfer (see `pvr_dest_addr()`). If you need to transfer to or from PVR RAM via the Root bus, then use DMA2, if you need to transfer to TA, then use TA DMA. Just needs some tweaking for`pvr_dma_transfer()` and that all.
And SQ's already can transfer to any memory, there is no point in copying their functions at all, they already do it by transfer `type`.

**Comment by andressbarajas** [comment]
@DC-SWAT Yep! Thats the goal. I just have two implementations to test things out and have the ability to compare. Another function besides pvr_dma_transfer that we have to think about is pvr_dma_ready().  Not sure how to handle that. I created this PR to get input but mainly to share with @gyrovorbis since he is way forward thinking when it comes to APIs. Im expecting him to finish the work I started.

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 1]
```diff
@@ -88,81 +145,137 @@ static uintptr_t pvr_dest_addr(uintptr_t dest, int type) {
...
+                            PVR_DMA_TO_PVR, callback, cbdata);
+}
+
+int pvr_dma_download_txr(void *sh4, pvr_ptr_t pvr, size_t count, int block,
+                    pvr_dma_callback_t callback, void *cbdata) {
+    return pvr_dma_transfer(sh4, (uintptr_t)pvr, count, PVR_DMA_VRAM64_SB, block, 
+                            PVR_DMA_TO_SH4, callback, cbdata);
+}
+
+int pvr_dma_load_pal(void *sh4, pvr_ptr_t pvr, size_t count, int block,
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:267]: Also can be applied to fog.

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 2]
```diff
@@ -20,224 +20,395 @@
...
 
     return dest_addr;
 }
 
-int pvr_dma_transfer(void *src, uintptr_t dest, size_t count, int type,
-                     int block, pvr_dma_callback_t callback, void *cbdata) {
-    uintptr_t src_addr = ((uintptr_t)src);
+int pvr_dma_transfer(void *src, void *dest, size_t count, pvr_dma_mode_t type,
+                    int block, pvr_dma_callback_t callback, void *cbdata) {
+    uintptr_t src_addr = (uintptr_t)src;
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:164]:
```suggestion
    uintptr_t src_addr = ((uintptr_t)src) & MEM_AREA_CACHE_MASK;
```

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 3]
```diff
@@ -20,224 +20,395 @@
...
+    uintptr_t dest_addr = (uintptr_t)dest;
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:165]:
```suggestion
    uintptr_t dest_addr = ((uintptr_t)dest) & MEM_AREA_CACHE_MASK;
```

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 4]
```diff
@@ -20,224 +20,395 @@
...
+
+        /* Make sure we're not already DMA'ing */
+        if(pvr_dma->start != 0) {
+            dbglog(DBG_ERROR, "pvr_dma: Previous DMA has not finished\n");
+            errno = EINPROGRESS;
+            return -1;
+        }
+
+        /* Figure out which direction we are going */
+        if(((dest_addr & MEM_AREA_CACHE_MASK) >> 24) == 0x0c) {
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:194]:
```suggestion
        if((dest_addr >> 24) == 0x0c) {
```

**Comment by andressbarajas** [comment]
@DC-SWAT @gyrovorbis Its READY! Looking forward to full review.

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 5]
```diff
@@ -20,224 +20,389 @@
...
+        pvr_dma->trigger = CPU_TRIGGER;
+
+        /* Start the DMA transfer */
+        pvr_dma->enable = 1;         
+        pvr_dma->start = 1;          
+
+        /* Block if necessary */
+        if(block)
+            sem_wait(&pvr_dma_done);
+    } else {
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:216]:
```suggestion
    }
    else {
```

Comment (DC-SWAT) [reply kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:216]: BlueCrab wouldn't like this.

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c [hunk 6]
```diff
@@ -20,224 +20,389 @@
...
+
+    *pvr_dma_pro = PVR_DMA_LOCK_ALLMEM;
+}
+
+static int check_dma_state(pvr_dma_mode_t type, const char *func_name) {
+    if(type >= PVR_DMA_VRAM32_SB && pvr_dma->start != 0) {
+        dbglog(DBG_ERROR, "%s: PVR DMA has not finished\n", func_name);
+        errno = EINPROGRESS;
+        return -1;
+    } else if(ta_dma->start != 0) {
```
Comment (DC-SWAT) [inline kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:363]:
```suggestion
    }
    else if(ta_dma->start != 0) {
```

**Review by DC-SWAT — APPROVED** [review]
Looks good.

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c (hunk 5 above)
Comment (andressbarajas) [reply kernel/arch/dreamcast/hardware/pvr/pvr_dma.c]: Done

File: kernel/arch/dreamcast/hardware/pvr/pvr_dma.c (hunk 6 above)
Comment (andressbarajas) [reply kernel/arch/dreamcast/hardware/pvr/pvr_dma.c]: Done

**Comment by andressbarajas** [comment]
> Is there any way we might avoid having `pvr_dma_load_txr` and `pvr_txr_load_dma`? I get the reasoning, but it seems like it would be nothing but confusing. Maybe by shifting the naming to emphasize 'rb' (root bus) for PVR and 'ta' for the TA then deprecating `pvr_txr_load_dma` entirely?
> 
So `rb_dma_load_txr()` and `ta_dma_load_txr()`

**Comment by andressbarajas** [comment]
@QuzarDC @gyrovorbis @pcercuei It might be a while before #539 gets merged so I fixed the conflicts and renamed the confusing load texture functions. Ready for review.

**Pull request closed without merging**